e2e/LongHaul/service/IotHub.cs

```diff
@@ -124,78 +75,56 @@ public async Task MonitorConnectedDevicesAsync(CancellationToken ct)
 
                 await foreach (ClientTwin device in allDevices)
                 {
-                    if (s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Disconnected)
+                    string deviceId = device.DeviceId;
+
+                    if (s_onlineDeviceCancellationTokenSources.ContainsKey(deviceId)
+                        && s_onlineDeviceOperations.ContainsKey(deviceId)
+                        && device.ConnectionState is ClientConnectionState.Disconnected)
                     {
-                        s_activeDeviceSet.Remove(device.DeviceId);
+                        CancellationTokenSource source = s_onlineDeviceCancellationTokenSources[deviceId];
+                        source.Cancel(); // Signal cancellation to all tasks on the particular device.
+                        s_onlineDeviceCancellationTokenSources.Remove(deviceId);
+                        s_onlineDeviceOperations.Remove(deviceId);
                     }
-                    else if (!s_activeDeviceSet.Contains(device.DeviceId) && device.ConnectionState == ClientConnectionState.Connected)
+                    else if (!s_onlineDeviceCancellationTokenSources.ContainsKey(deviceId)
+                        && !s_onlineDeviceOperations.ContainsKey(deviceId)
+                        && device.ConnectionState is ClientConnectionState.Connected)
                     {
-                        s_activeDeviceSet.Add(device.DeviceId);
+                        // For each online device, initiate a new cancellation token source.
+                        // Once the device goes offline, cancel all operations on this device.
+                        var source = new CancellationTokenSource();
+                        CancellationToken token = source.Token;
+                        s_onlineDeviceCancellationTokenSources.Add(deviceId, source);
+
+                        s_onlineDeviceOperations.Add(
```

Comment:
I think this would read better if you instead got a task from this and then put it in the dictionary on line 118. The way it is requires both actions to access the dictionary, which is an opportunity for a bug.